cranelift-codegen/src/verifier/mod.rs

```diff
@@ -276,6 +289,47 @@ impl<'a> Verifier<'a> {
         }
     }
 
+    /// Build the context of an instruction for better errors.
+    #[inline]
+    fn context(&self, inst: Inst) -> Option<String> {
+        Some(self.func.dfg.display_inst(inst, self.isa).to_string())
+    }
+
+    /// Helper for appending a contextual error to the verifier error list.
+    #[inline]
+    fn append_error(&self, inst: Inst, errors: &mut VerifierErrors, message: impl Into<String>) {
+        let location = inst.into();
+        let context = self.context(inst);
+        let message = message.into();
+        errors.0.push(VerifierError {
+            location,
+            context,
+            message,
+        });
+    }
+
+    /// Diagnose a fatal error, and return `Err`.
+    fn fatal(
```

Comment:
Nice! Can the `fatal` and `nonfatal` macros be entirely removed now, by replacing other uses in the other files contained in the `verifier` directory?

Reply:
`fatal!` still has 52 usages, `nonfatal!` has 4 usages, and `report!` has 23 usages; I'll create an issue to do that as a separate PR because I think I would like to refactor that stuff on to `VerifierErrors`: #1248.